### DISCUSSION TITLE
Example of adding Baggage to a Trace

### DISCUSSION THREAD

**Discussion by daschneider** [open-telemetry/opentelemetry-java#4801] [Q&A]
I need to add some domain specific baggage values to our traces.  I've reviewed the Javadoc and examples and I still don't see how to go about doing this.  Can anyone point me to a working example of how to add baggage to a trace?

Thanks in advance.

**Comment by jack-berg** [comment]
You can create a span inside of a scope which has had the baggage added to it, like so:
```
    try (var baggageScope = Baggage.builder()
        .put("user.name", "jack")
        .build()
        .storeInContext(Context.current()).makeCurrent()) {
      var span = TRACER.spanBuilder("my-span").startSpan();
      try (var scope = span.makeCurrent()) {
        // do work
      } finally {
        span.end();
      }
    }
```
Then you can add a SpanProcessor to your SdkTracerProvider config which extracts the baggage and appends as span attributes. Here's an example which appends all baggage keys as span attributes:
```
    SdkTracerProvider.builder()
            .addSpanProcessor(new SpanProcessor() {
              @Override
              public void onStart(Context parentContext, ReadWriteSpan span) {
                Baggage.current().asMap().forEach((s, baggageEntry) -> span.setAttribute(s, baggageEntry.getValue()));
              }

              @Override
              public boolean isStartRequired() {
                return true;
              }

              @Override
              public void onEnd(ReadableSpan span) {
              }

              @Override
              public boolean isEndRequired() {
                return false;
              }
            })
```

**Reply by daschneider** [reply]
Thanks for the reply @jack-berg.  That doesn't appear to be my use case.  I'm not asking how to add baggage values to a span as attributes, I'm asking how to add new baggage values to an existing trace.

**Reply by jack-berg** [reply]
Although baggage is intended to be useful for tracing, the baggage API exists separately from the trace API. I.e. you can use baggage without using the trace API at all. Baggage is just another value that can be stored inside of Context. As such, you can create a add entries to baggage, store the baggage in context, and setup a scope for that context:
```
    try (var baggageScope = Baggage.builder()
        .put("user.name", "jack")
        .build()
        .storeInContext(Context.current()).makeCurrent()) {
        // do work in the scope of the baggage
    }
```
Note: you cannot update baggage in the current context - you can only change it and create a new context.

Any work that's done in the scope of the baggage will be able to access it. From a tracing perspective, you might want to extract baggage entries and include them on spans. I've already shown how to do that. You might also wish to propagate baggage across service boundaries. If this is the case, you can configure the SdkTracerProvider to use the W3CBaggagePropagator.

**Reply by pavel-luhin** [reply]
Hi @jack-berg. So how can I add a value to the baggage and access it in another service?
I saw that in brave I can just autowire `BaggageField` and set it value, but how can I do it in OpenTelemetry?

I see your response how to add field to the baggage and create new context, but this context won't be propagated later in other service.

**Reply by jkwatson** [reply]
You will need to propagate that context, by making that context "current" for whatever code executes. Jack's example spells that out, as the `// do work in the scope of the baggage` is where the downstream code needs to live in order for the baggage to be propagated.

**Reply by pavel-luhin** [reply]
@jkwatson Hello! Thank you for the answer. Is there any other way to propagate new baggage fields? I need to do this in `spring-cloud-gateway` filter and seems it does not work well with its filter chain.

```
try (var baggageScope = Baggage.builder()
        .put("user.name", "jack")
        .build()
        .storeInContext(Context.current()).makeCurrent()) {
        // do work in the scope of the baggage
        return chain.filter(updatedExchange)
}
```

And using debug in `W3CBaggagePropagator` I see that baggage is empty. In configuration I set up propagation type to `W3C`, so the problem is that context is changed. Also consuming service does not receive baggage header.

**Reply by jkwatson** [reply]
This code doesn't look right to me. `storeInContext` returns a context, not a scope.  I think this code should look something like this:

```
var baggage = Baggage.builder().put("user.name", "jack").build();
try (Scope ignored = baggage.makeCurrent()) {
  return chain.filter(updatedExchange);
}
```

**Reply by pavel-luhin** [reply]
@jkwatson Hello.
I am doing `storeInContext().makeCurrent()`, so `makeCurrent` returns `Scope`. 
The code that you provided does not work too. I am thinking, maybe the reason is because I am using Kotlin and its function `.use`.

```
let baggage = Baggage.builder()
            .put("enable-extra", enableExtra)
            .build()

baggage.makeCurrent().use {
            return chain.filter(updatedExchange)
}
```

In debug I see that even on 1 level higher after return `Context.current()` has empty baggage.
But in `.use` function `Context.current()` has baggage with `enable-extra` field.

**Reply by jkwatson** [reply]
I don't know what you mean by "1 level higher" in this context. The current Context will only have the new value in it until it falls out of scope, so if it has the value inside your `use` function, then it is working as it should.
